src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,125 @@
+        }
+
+        for match in self.matches:
+            try:
+                lintable = lintables[match.filename]
+            except KeyError:
+                # we shouldn't get here, but this is easy to recover from so do that.
+                lintable = Lintable(match.filename)
+                self.matches_per_file[lintable] = []
+            self.matches_per_file[lintable].append(match)
+
+    def run(self) -> None:
+        """Execute the fmt transforms."""
+        # ruamel.yaml rt=round trip (preserves comments while allowing for modification)
+        yaml = YAML(typ="rt")
+
+        # NB: ruamel.yaml does not have typehints, so mypy complains about everything here.
+
+        # configure yaml dump formatting
+        yaml.explicit_start = (  # we handle adding --- in self._final_yaml_transform()
+            False  # type: ignore[assignment]
+        )
+        yaml.explicit_end = False  # type: ignore[assignment]
+
+        # TODO: make the width configurable
+        # yaml.width defaults to 80 which wraps longer lines in tests
+        yaml.width = 120  # type: ignore[assignment]
+
+        yaml.default_flow_style = False
+        yaml.compact_seq_seq = (  # dash after dash
+            True  # type: ignore[assignment]
+        )
+        yaml.compact_seq_map = (  # key after dash
+            True  # type: ignore[assignment]
+        )
+        # yaml.indent() obscures the purpose of these vars:
+        yaml.map_indent = 2  # type: ignore[assignment]
+        yaml.sequence_indent = 4 if self.indent_sequences else 2  # type: ignore[assignment]
+        yaml.sequence_dash_offset = yaml.sequence_indent - 2  # type: ignore[operator]
+
+        # explicit_start=True + map_indent=2 + sequence_indent=2 + sequence_dash_offset=0
+        # ---
+        # - name: playbook
+        #   loop:
+        #   - item1
+        #
+        # explicit_start=True + map_indent=2 + sequence_indent=4 + sequence_dash_offset=2
+        # ---
+        #   - name: playbook
+        #     loop:
+        #       - item1
+
+        for file, matches in self.matches_per_file.items():
+            # str() convinces mypy that "text/yaml" is a valid Literal.
+            # Otherwise, it thinks base_kind is one of playbook, meta, tasks, ...
+            file_is_yaml = str(file.base_kind) == "text/yaml"
+
+            if file_is_yaml:
+                # load_data has an lru_cache, so using it should be cached vs using YAML().load() to reload
+                data: Union[CommentedMap, CommentedSeq] = load_data(file.content)
+                yaml.dump(data, file.path, transform=self._final_yaml_transform)
+
+    def _final_yaml_transform(self, text: str) -> str:
+        """
+        Ensure that top-level sequences are not over-indented.
+
+        In order to make that work, we cannot delegate adding the yaml explict_start
+        to ruamel.yaml or dedent() won't have anything to work with.
+        Instead, we add the explicit_start here.
+        """
```

Comment:
I just ran `ansible-lint --fmt-yaml` on a few thousand yaml files... And this function did not do very well. I ran into a variety of edge cases with comments and one case where a yaml block (`|`) had a line that started with `#`, so we got a false positive on comment handling.

So, this approach doesn't work well enough for widespread use.
But I found a way to delegate this formatting to ruamel.yaml! Using a subclass of `ruamel.yaml.emitter.Emitter` I was able to customize the sequence indents so that it does not indent root-level sequences.

I'll clean that up and then push it later.

Reply:
OK. I replaced this function with a `FormattedEmitter` class.